src/runner.py

```diff
@@ -88,596 +47,416 @@ class LxdInstanceConfig(TypedDict):
-        """
-        # Redundant None check for mypy type check.
-        return self.is_online and self.remote is not None and self.remote["busy"] is False
+        self.name = name
+        self.exist = exist
+        self.online = online
+        self.busy = busy
 
-    @property
-    def is_offline(self) -> bool:
-        """Whether the runner is offline.
+        self.instance = instance
 
-        Returns:
-            Whether the runner is offline.
-        """
-        return self.remote is not None and self.remote["status"] != "online"
+    def create(
+        self,
+        image: str,
+        resources: VirtualMachineResources,
+        binary: Path,
+        reconcile_interval: int,
+        registration_token: str,
+    ):
+        logger.info("Creating runner: %s", self.name)
 
-    @property
-    def is_online(self) -> bool:
-        """Whether the runner is online.
+        self.instance = self._create_instance(image, resources)
 
-        Returns:
-            Whether the runner is online.
-        """
-        return self.remote is not None and self.remote["status"] == "online"
+        try:
+            self._start_instance(reconcile_interval)
+            self._install_binary(binary)
+            self._configure_runner()
+            self._register_runner(registration_token, labels=[self.app_name, image])
+            self._start_runner()
+        except Exception as err:
+            self.instance.stop(wait=True)
+            try:
+                self.instance.delete(wait=True)
+            except Exception:  # nosec B110
+                # this is just a fall-back.
+                # Ephemeral containers should auto-delete when stopped;
+                pass
+            raise RunnerCreateError(str(err)) from err
 
-    @property
-    def is_local(self) -> bool:
-        """Whether the runner is local.
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5, logger=logger)
+    def remove(self) -> None:
+        """Remove this runner instance from LXD and GitHub.
 
-        Returns:
-            Whether the runner is local.
+        Raises:
+            RunnerRemoveError: Failure in removing runner.
         """
-        return self.local is not None
-
-    @property
-    def virt_type(self) -> Optional[str]:
-        """Return the virtualization type of the runner.
 
-        Returns:
-            Virtualization type of the runner.
-        """
-        if self.is_local:
-            # Assert for mypy type check. This check is already done by `is_local`.
-            assert self.local is not None  # nosec B101
-            return self.local.type
-        elif self.remote:
-            remote_virt_type = (
-                label["name"]
-                for label in self.remote["labels"]
-                if label["name"] in {"container", "virtual-machine"}
-            )
-            return next(remote_virt_type, None)
-        return None
+        @retry(tries=10, delay=10, logger=logger)
```

Comment:
All the retry within retry are for function that are waiting for something to succeed before moving on.
E.g., booting up an instance.

After some thought, rather than retry within retry it should be in sequence.
E.g., booting up with retries, then verify successfully boot up with retries.
So stacking retries are not an problem.